### DISCUSSION TITLE
Running Xorg as a service.

### DISCUSSION THREAD

**Discussion by SolitudeSF** [davmac314/dinit#218] [Q&A] [answered]
Im trying to run Xorg as a user service. Practicality of this in my current setup, i admit, is dubious, but i have an itch to make it work. Im running relatively recent `dinit`, launching it from `.profile` instead of `startx`.
I have this as a definition of `xorg` service:
```
type = process
command = $XDG_CONFIG_HOME/dinit/scripts/startx
ready-notification = pipevar:DINIT_READYFD
options = runs-on-console
restart = no
```

and this is the launcher script (stripped down `sx`):
```sh
#!/bin/sh --
tty=$(tty)
tty=${tty#/dev/tty}
dinitctl setenv DISPLAY=:"$tty"

cfgdir=${XDG_CONFIG_HOME:-$HOME/.config}/sx
datadir=${XDG_DATA_HOME:-$HOME/.local/share}/sx
mkdir -p -- "$cfgdir" "$datadir"

export XAUTHORITY="${XAUTHORITY:-$datadir/xauthority}"
touch -- "$XAUTHORITY"

xauth add :"$tty" MIT-MAGIC-COOKIE-1 "$(od -An -N16 -tx /dev/urandom | tr -d ' ')"

Xorg :"$tty" -keeptty vt"$tty" -noreset -auth "$XAUTHORITY" -displayfd "$DINIT_READYFD"
```

the issue is that when `xorg` runs, Xorg successfully starts and after going through initialization instantly stops with `Server succesfully terminated (0)`, without any error in sight. When the same script is ran by hand from a tty, Xorg launches and stays running as expected. Am i missing some configuration option? What is the difference between the environments?

**Comment by mobin-2008** [comment]
Hello. Take a look at `~/.local/share/xorg/` log files both from login shell and from dinit.

**Reply by davmac314** [reply]
The service description has no `logfile =` entry, so there won't be any log from dinit.

**Reply by davmac314** [reply]
I think what you mean is: compare the Xorg log file from `~/.local/share/xorg/` (it will be named `Xorg.0.log` for example) for when X is started successfully (via the shell) versus when it is started unsuccessfully by Dinit. (That isn't very clear from the message.)

**Comment by davmac314** [comment]
One obvious problem that I can see is the `-displayfd "$DINIT_READYFD"` argument to Xorg. The readiness fd is for readiness notification, this appears to be telling Xorg to use it for the display (which isn't possible).

**Reply by SolitudeSF** [reply]
`-displayfd fd          file descriptor to write display number to when ready to connect`

**Reply by davmac314** [reply]
Ah you're totally right, my bad.
Did you check the Xorg log as mobin suggested?

**Reply by davmac314** [reply]
<strike>(Also might be worth setting `logfile =` entry in the service description and see if you get any output to it).</strike>

Sorry, forgot it was `runs-on-console` so this isn't appropriate. Not thinking straight today.

**Reply by davmac314** [reply]
One other thing: is dinit actually connected to the tty? A `runs-on-console` service just gets dinit's own stdin/stdout, so if they are not the terminal, then the `tty=$(tty)` line will probably fail (with "not a tty").

**Comment by SolitudeSF** [comment]
One thing i missed is `setenv`ing `XAUTHORITY` too, so dependant services can use Xorg. After `enabling`/`disabling` `xorg` service few times and some other misc rearrangements it no longer immediately stops, and i dont have patience to try reproducing it. So it works now.

**Reply by SolitudeSF** [reply, marked as answer]
Computers are magic.
